### ISSUE TITLE
risky unit tests "Test code or tested code did not (only) close its own output buffers"

### ISSUE THREAD

**Issue by staabm** [filp/whoops#534]
taken from a recent travis ci build

```
1) Whoops\Handler\PlainTextHandlerTest::testReturnsWithFramesAndArgsOutput

Test code or tested code did not (only) close its own output buffers

2) Whoops\Handler\PlainTextHandlerTest::testReturnsWithLoggerOnlyOutput

Test code or tested code did not (only) close its own output buffers
```

https://travis-ci.org/filp/whoops/jobs/283099825

we should check what we can do about that.

**Comment by Magentron** [comment]
I am experiencing this error as well, and it was caused by an exception thrown in `Woops\Run::handleException()` by the call to `$handler->handle($exception);` after processing a controller action that throws an `AccessDeniedHttpException` with a custom message.

The exception thrown in `handle()` is an ErrorException because of the error `file_get_contents(Standard input code): failed to open stream: No such file or directory`, which seems to be caused by running the test in a separate process:
```
 * @runTestsInSeparateProcesses
 * @preserveGlobalState disabled
 * @backupGlobals disabled
``` 

Except for the unexpected output buffering status, the test runs successfully and if an empty exception catch is put around the call to `handle()`, the test completes successfully.

It seems the problems is caused by missing exception handling in `handleException()` in `Whoops\Run.php` which seems to be solved by adding a **finally {}** block where the output is obtained, like so:

```
        try {
            foreach (array_reverse($this->handlerStack) as $handler) {
                $handler->setRun($this);
                $handler->setInspector($inspector);
                $handler->setException($exception);
    
                // The HandlerInterface does not require an Exception passed to handle()
                // and neither of our bundled handlers use it.
                // However, 3rd party handlers may have already relied on this parameter,
                // and removing it would be possibly breaking for users.
                $handlerResponse = $handler->handle($exception);
    
                // Collect the content type for possible sending in the headers.
                $handlerContentType = method_exists($handler, 'contentType') ? $handler->contentType() : null;
    
                if (in_array($handlerResponse, [Handler::LAST_HANDLER, Handler::QUIT])) {
                    // The Handler has handled the exception in some way, and
                    // wishes to quit execution (Handler::QUIT), or skip any
                    // other handlers (Handler::LAST_HANDLER). If $this->allowQuit
                    // is false, Handler::QUIT behaves like Handler::LAST_HANDLER
                    break;
                }
            }
    
            $willQuit = $handlerResponse == Handler::QUIT && $this->allowQuit();
        } catch (\Throwable $e) {
            throw $e;
        } finally {
            $output = $this->system->cleanOutputBuffer();
        }
```

Since I'm not too familiar with this framework there's probably a more elegant solution.
Please let me know your thoughts.

**Comment by staabm** [comment]
sounds plausible. I have created a PR which implements your idea:

https://github.com/filp/whoops/pull/632

**Comment by denis-sokolov** [comment]
Seems we’ve closed this in #632.
